up-l1/README.adoc

```diff
@@ -50,157 +42,218 @@ We will use the following terminology during this document:
+The listener will no longer be called for any (matching) messages after this method has returned successfully.
 
-`OUT <<UStatus>> unregisterListener(IN UUri, IN <<UListener>>&)`
+[source]
+----
+unregisterListener(
+  IN sourceFilter: UUri,
+  IN sinkFilter: UUri?,
+  IN listener: UListener&,
+  OUT result: Future<Void, UStatus>
+)
+----
 
-==== Parameters
-.UnregisterListener Parameters
-[width="70%",cols="20%,80%"]
+.RegisterListener Parameters
+[width="100%",cols="15%,25%,60%"]
 |===
-|Parameter | Description
+|Parameter | Type | Description
 
-| UUri
-| Topic to unregister the listener for
+| sourceFilter
+| link:../basics/uri.adoc[UUri]
+| The source address pattern that the listener had been registered for.
 
+| sinkFilter
+| link:../basics/uri.adoc[UUri]
+| The sink address pattern that the listener had been registered for.
+
+| listener
 | <<UListener>>
-| Listener to be unregistered
+| The listener to be unregistered.
 
+| result
+| Future<Void, link:../basics/error_model.adoc[UStatus]>
+| A succeeded future if the listener has been unregistered successfully. +
+Otherwise, a failed future containing a `UStatus` indicating the reason for the failure.
 |===
 
-* *MUST* pass a UUri for which `isMicroForm()` returns true
-* *SHOULD* pass a Resolved UUri (contains both names and numbers)
-
 
-=== Receive()
+=== Receive
 
-Implements the _pull_ <<Delivery Method>> to fetch a message from the transport for a given UUri (topic).
+Clients use this method to receive a single message matching given filter criteria.
+This method implements the _pull_ <<Delivery Method>> for the underlying messaging infrastructure.
 
-`OUT <<UStatus>> receive(IN UUri, OUT UPayload&, OUT UAttributes&)`
+[source]
+----
+receive(
+  IN sourceFilter: UUri,
+  IN sinkFilter: UUri?,
+  OUT result: Future<UMessage&, UStatus>
+)
+----
 
-==== Parameters
 .Receive Parameters
-[width="100%",cols="20%,80%"]
+[width="100%",cols="15%,25%,60%"]
 |===
-|Parameter | Description
-
-| UUri
-| Topic to receive the message from
-
-| UPayload
-a| Data and metadata received
+|Parameter | Type | Description
 
- * *MUST* be passed by reference
+| sourceFilter
+| link:../basics/uri.adoc[UUri]
+| The _source_ address pattern that messages need to match.
 
-| UAttributes
-a| Message metadata
+| sinkFilter
+| link:../basics/uri.adoc[UUri]
+| The _sink_ address pattern that messages need to match. If omitted, a message must not contain any sink address in order to match.
 
- * *MUST* be passed by reference
+| result
+| Future<Void, link:../basics/error_model.adoc[UStatus]>
+| A succeeded future containing the least recent message that matches the given filter criteria and has not expired yet. +
+A failed future if no message can be retrieved from the messaging infrastructure. The `UStatus` will contain a `UCode` indicating the reason for the failure.
 
 |===
 
- * *MUST* return `NOT_FOUND` if there are no messages for the given topic
+[oft-sid="req~utransport-register-listener~1"]
```

Comment:
I would like to discuss adding this metadata to the rest of the specs, was this one added to test out the links and then if we agree we would add them to all requirements? How do we link this in our code?

Reply:
Adding these might indeed be a little premature. Ideally, they would serve as OpenFastTrace specification item IDs which could then be included in source code, tests etc to provide for traceability across all artifacts. I haven't yet tested this, I have to admit. In fact, OFT does not (yet) support AsciiDoc files. However, from what I can tell, it should not be too hard to add this to OpenFastTrace.

Reply:
I would love that we can do this one a single file (ex. the uri.adoc) and see if this can work end-2-end before we start adding the tags everywhere and then have to remove/change them later.

Reply:
Yes, that is what I have in mind as well. I have spent some time investigating OpenFastTrace's extension API today. Looks very doable ...